### PR TITLE
[PLAT-1664] Adding UX Loading improvements to the scene tree

### PR DESCRIPTION
**Pull request by MadisonEhlers-Vertex**
## Summary

These changes support adding a loader to the search bar, as well as a loader to the scene tree when items are taking beyond `2s`

loading

### CODE REVIEW

**Review by amvertex — APPROVED** [review]

File: packages/viewer/src/components/scene-tree/scene-tree.spec.tsx [hunk 1]
```diff
@@ -802,7 +802,6 @@ describe('<vertex-scene-tree>', () => {
       const row2 = await tree.getRowAtIndex(2);
       await tree.selectItem(row2, { range: true });
 
-      expect(true).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (sourada-e5) [reply packages/viewer/src/components/scene-tree/scene-tree.spec.tsx:805]: The day that fails... run for the hills.

File: packages/viewer/src/components/scene-tree-table-layout/scene-tree-table-layout.tsx [hunk 2]
```diff
@@ -258,6 +265,23 @@ export class SceneTreeTableLayout {
...
+        this.setIsLoading,
+        DISPLAY_LOADER_TIMEOUT
+      );
+    } else if (isViewportDataPresent) {
+      this.isStillLoading = false;
+      if (this.loadingTimer) {
+        clearTimeout(this.loadingTimer);
+      }
+      this.loadingTimer = undefined;
+    }
```
Comment (danschultz) [inline packages/viewer/src/components/scene-tree-table-layout/scene-tree-table-layout.tsx:284]: This is pretty fragile. We're using the rendering mechanics of Stencil to infer loading status..

What if we added a `loading` field to `SceneTreeState` and have the `SceneTreeController` manage this field. This way we can bind the view to the current state.

File: packages/viewer/src/components/scene-tree-table-layout/scene-tree-table-layout.tsx [hunk 3]
```diff
@@ -334,10 +358,21 @@ export class SceneTreeTableLayout {
           <slot name="divider" />
         </div>
         {this.resizingColumnIndex != null && <div class="resize-overlay" />}
+        {this.isStillLoading && (
+          <slot name="loading">
+            <div class="loading">
+              <vertex-viewer-spinner size="md" />
+            </div>
+          </slot>
+        )}
```
Comment (danschultz) [inline packages/viewer/src/components/scene-tree-table-layout/scene-tree-table-layout.tsx:367]: This component should be responsible for the layout of rows. I think the responsibility of displaying loading spinner is more appropriate in `vertex-scene-tree`.

File: packages/viewer/src/components/scene-tree/scene-tree.tsx [hunk 4]
```diff
@@ -838,16 +841,18 @@ export class SceneTree {
...
   @Listen('search')
-  protected handleSearch(event: CustomEvent<string>): void {
+  protected async handleSearch(event: CustomEvent<string>): Promise<void> {
     const columnsToSearch =
       this.metadataSearchKeys.length > 0
         ? this.metadataSearchKeys
         : this.metadataKeys;
 
-    this.filterItems(event.detail, {
+    this.isSearching = true;
```
Comment (danschultz) [inline packages/viewer/src/components/scene-tree/scene-tree.tsx:850]: I'd argue for putting this field in `SceneTreeState` as well. This way the controller is the central place for managing this state. Right now, I bet there's a bug because someone can either call `SceneTree.filter()` or `SceneTreeController.filter()` and you won't get a loading spinner.

File: packages/viewer/src/components/viewer-spinner/viewer-spinner.css [hunk 5]
```diff
@@ -0,0 +1,94 @@
+
+
```
Comment (danschultz) [inline packages/viewer/src/components/viewer-spinner/viewer-spinner.css:2]: Unnecessary whitespace.

File: packages/viewer/src/components/viewer-spinner/viewer-spinner.css [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+.spinner {
+  position: relative;
+  display: flex;
```
Comment (danschultz) [inline packages/viewer/src/components/viewer-spinner/viewer-spinner.css:5]: Does this need to be a flex container?

File: packages/viewer/src/components/viewer-spinner/viewer-spinner.css [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+.spinner div {
+  position: absolute;
+  display: block;
+  animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
+  border-radius: 50%;
+  border-style: solid;
+  box-sizing: border-box;
+  border-color: var(--neutral-800) transparent transparent transparent;
+  border-width: var(--border-width, 1px);
```
Comment (danschultz) [inline packages/viewer/src/components/viewer-spinner/viewer-spinner.css:18]: This CSS prop is not documented as part of the component. Also, please use the CSS prop naming conventions like the other components.

File: packages/viewer/src/components/viewer-spinner/viewer-spinner.tsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+   *  * `md`: 32px
+   *  * `lg`: 64px
+   */
+  @Prop()
+  public size?: SpinnerSize = 'md';
+
+  public render(): h.JSX.IntrinsicElements {
+    return (
+      <div
+        class={classNames('spinner grey', {
```
Comment (danschultz) [inline packages/viewer/src/components/viewer-spinner/viewer-spinner.tsx:30]: `grey` is not used anywhere.

File: packages/viewer/src/components/viewer-spinner/viewer-spinner.css [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  height: 16px;
+  width: 16px;
+  margin: 0;
+}
+
+/* sm */
+.spinner.sm {
+  height: 23px;
+  width: 23px;
```
Comment (danschultz) [inline packages/viewer/src/components/viewer-spinner/viewer-spinner.css:51]: This is an unusually odd size in web dev. Why not 24px, which is the size of the child div `18px + 2px margin/side`.

File: packages/viewer/src/components/viewer-spinner/viewer-spinner.css [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+/* lg */
+.spinner.lg {
+  height: 64px;
+  width: 64px;
+}
+
+.spinner.lg div {
+  border-width: var(--border-width, 6px);
+  height: 51px;
+  width: 51px;
```
Comment (danschultz) [inline packages/viewer/src/components/viewer-spinner/viewer-spinner.css:83]: `52px`? This'd equal the size defined in `.spinner.lg`.
